Index: linux-4.4.38/arch/mips/ralink/rt305x.c
===================================================================
--- linux-4.4.38.orig/arch/mips/ralink/rt305x.c
+++ linux-4.4.38/arch/mips/ralink/rt305x.c
@@ -48,7 +48,10 @@ static struct rt2880_pmx_func rgmii_func
 static struct rt2880_pmx_func rt3352_lna_func[] = { FUNC("lna", 0, 36, 2) };
 static struct rt2880_pmx_func rt3352_pa_func[] = { FUNC("pa", 0, 38, 2) };
 static struct rt2880_pmx_func rt3352_led_func[] = { FUNC("led", 0, 40, 5) };
-
+static struct rt2880_pmx_func rt3352_cs1_func[] = {
+	FUNC("spi_cs1", 0, 45, 1),
+	FUNC("wdg_cs1", 1, 45, 1),
+};
 static struct rt2880_pmx_group rt3050_pinmux_data[] = {
 	GRP("i2c", i2c_func, 1, RT305X_GPIO_MODE_I2C),
 	GRP("spi", spi_func, 1, RT305X_GPIO_MODE_SPI),
@@ -74,6 +77,7 @@ static struct rt2880_pmx_group rt3352_pi
 	GRP("lna", rt3352_lna_func, 1, RT3352_GPIO_MODE_LNA),
 	GRP("pa", rt3352_pa_func, 1, RT3352_GPIO_MODE_PA),
 	GRP("led", rt3352_led_func, 1, RT5350_GPIO_MODE_PHY_LED),
+	GRP("spi_cs1", rt3352_cs1_func, 2, RT3352_GPIO_MODE_SPI_CS1),
 	{ 0 }
 };
 
Index: linux-4.4.38/arch/mips/include/asm/mach-ralink/rt305x.h
===================================================================
--- linux-4.4.38.orig/arch/mips/include/asm/mach-ralink/rt305x.h
+++ linux-4.4.38/arch/mips/include/asm/mach-ralink/rt305x.h
@@ -139,6 +139,7 @@ static inline int soc_is_rt5350(void)
 #define RT5350_GPIO_MODE_SPI_CS1	21
 #define RT3352_GPIO_MODE_LNA		18
 #define RT3352_GPIO_MODE_PA		20
+#define RT3352_GPIO_MODE_SPI_CS1	21
 
 #define RT3352_SYSC_REG_SYSCFG0		0x010
 #define RT3352_SYSC_REG_SYSCFG1         0x014
